plugins/utils/navbuddy.nix

```diff
@@ -0,0 +1,267 @@
+            border = helpers.defaultNullOpts.mkBorder "rounded" "double" "solid" "none" null ''
+              "rounded", "double", "solid", "none"  or an array with eight chars building up the border in a clockwise fashion
+              starting with the top-left corner. eg: { "╔", "═" ,"╗", "║", "╝", "═", "╚", "║" }.
+            '';
+          };
+
+          right = {
+            border = helpers.defaultNullOpts.mkBorder "rounded" "double" "solid" "none" null ''
+              "rounded", "double", "solid", "none"  or an array with eight chars building up the border in a clockwise fashion
+              starting with the top-left corner. eg: { "╔", "═" ,"╗", "║", "╝", "═", "╚", "║" }.
+            '';
+            preview = helpers.defaultNullOpts.mkEnum [ "leaf" "always" "never" ] "leaf" ''
+              Right section can show previews too.
+                    Options: "leaf", "always" or "never"
+            '';
+          };
+        };
+      };
+
+      nodeMarkers = {
+        enabled = true;
+        icons = {
+          leaf = helpers.defaultNullOpts.mkStr "  " ''
+            The icon to use for leaf nodes.
+          '';
+
+          leaf_selected = helpers.defaultNullOpts.mkStr " → " ''
+            The icon to use for selected leaf node.
+          '';
+          branch = helpers.defaultNullOpts.mkStr "  " ''
+            The icon to use for branch nodes.
+          '';
+        };
+      };
+
+      icons =
+        mapAttrs
+          (
+            name: default:
+              default "icon for ${name}."
+          )
+          {
+            File = "󰈙 ";
+            Module = " ";
+            Namespace = "󰌗 ";
+            Package = " ";
+            Class = "󰌗 ";
+            Method = "󰆧 ";
+            Property = " ";
+            Field = " ";
+            Constructor = " ";
+            Enum = "󰕘";
+            Interface = "󰕘";
+            Function = "󰊕 ";
+            Variable = "󰆧 ";
+            Constant = "󰏿 ";
+            String = "󰀬 ";
+            Number = "󰎠 ";
+            Boolean = "◩ ";
+            Array = "󰅪 ";
+            Object = "󰅩 ";
+            Key = "󰌋 ";
+            Null = "󰟢 ";
+            EnumMember = " ";
+            Struct = "󰌗 ";
+            Event = " ";
+            Operator = "󰆕 ";
+            TypeParameter = "󰊄 ";
+          };
+
+      useDefaultMapping = helpers.defaultNullOpts.mkBool true ''
+        If set to false, only mappings set by user are set. Else default mappings are used for keys that are not set by user
+      '';
+
+      mapping = mkOption {
+        type = types.attrsOf types.str;
+        description = ''
+          Actions to be triggered for specified keybindings. 
+          For each keybinding it takes a table of format { callback = <function_to_be_called>, description = "string"}.
+          The callback function takes the "display" object as an argument.
+        '';
+        default = {
+          "<esc>" = "close";
```

Comment:
Do we need to set a description here, feels like we should ? If so what typing should I use ? `types.attrsOf types.str;` won't workl I guess?

Reply:
Something like we have in `trouble.nix` ? https://github.com/nix-community/nixvim/blob/32a64af2318beac3c89d4d1fa0b73ce237c5cf1e/plugins/lsp/trouble.nix#L51-L64

Reply:
I was more thinking of mapping keys to actions as a user could want two bindings to the same action.

Reply:
This is what you have done here in the option definition.
However, you need to adapt the implementation down below.

Reply:
Sorry I dont follow, does the mapping here need to change ? 

```nix
    mappings =
      mapAttrs'
      (action: key: {
        name = key;
        value = {
          action = "function() require('nvim-navbuddy.actions').actions.${action}) end";
          lua = true;
          inherit (cfg.keymaps) silent;
          #desc = "Spider-${motion}";
        };
      })
```